File: Sources/LoafCat/Modules/HuntModule.swift
import AppKit

/// Stalking a cursor that is being wiggled like a cat toy.
///
/// The thing this deliberately is NOT is a speed threshold. A speed threshold fires
/// on every ordinary sweep across a wide display, which is most cursor movement, and
/// then the cat is permanently crouched. What actually reads as "prey" is *changing
/// direction*: a real cat ignores a mouse running in a straight line past it and
/// fixates on one that jinks.
///
/// So this is an energy accumulator whose reversal term dwarfs its speed term. The
/// speed term exists only to gate — its ceiling, `gain * excess / (1 - decay)`, is
/// tuned to sit below the trigger for any plausible straight sweep, so speed alone
/// can never pounce the cat. Reversals are what carry it over.
final class HuntModule: CatModule, AtlasTuned {
    let id = "hunt"
    var tunedGeneration = -1

    // --- constants, every one of them from cat.json -------------------------
    private var decayPerFrame: CGFloat = 1
    private var speedMin: CGFloat = 0
    private var speedGain: CGFloat = 0
    private var reverseLag: Double = 0
    private var reverseDot: CGFloat = 0
    private var reverseSpeed: CGFloat = 0
    private var reverseBonus: CGFloat = 0
    private var refractory: Double = 0
    private var accelMin: CGFloat = 0
    private var accelGain: CGFloat = 0
    private var trigger: CGFloat = 1
    private var resetTo: CGFloat = 0
    private var crouchTime: Double = 0
    private var recoverTime: Double = 0
    private var attack: CGFloat = 0.1
    private var crouchSquash: CGFloat = 1
    private var lean: CGFloat = 0
    private var bodyLean: CGFloat = 0
    private var pawReach: CGFloat = 0
    private var wiggleHz: CGFloat = 0
    private var wiggleAmp: CGFloat = 0

    // --- state --------------------------------------------------------------
    private enum Phase { case idle, crouch, recover }
    private var phase: Phase = .idle
    private var phaseUntil: CFAbsoluteTime = 0
    private var energy: CGFloat = 0
    private var pose: CGFloat = 0
    private var elapsed: CGFloat = 0
    private var lastVelocity = CGPoint.zero
    private var lastReverse: CFAbsoluteTime = 0
    private var reversals: CGFloat = 0

    /// A short trail of velocities, so a reversal is measured against where the
    /// cursor was heading a moment ago rather than one frame ago. At 120Hz the
    /// frame-to-frame angle is mostly EMA noise; over ~60ms it is the gesture.
    private var trail: [(t: CFAbsoluteTime, v: CGPoint)] = []
    private let trailCap = 48

    func retune(_ atlas: Atlas) {
        let b = atlas.behaviour
        decayPerFrame = b.f("hunt.decay_per_frame")
        speedMin = b.f("hunt.speed_min")
        speedGain = b.f("hunt.speed_gain_per_frame")
        reverseLag = Double(b.f("hunt.reverse_lag"))
        reverseDot = b.f("hunt.reverse_dot")
        reverseSpeed = b.f("hunt.reverse_speed")
        reverseBonus = b.f("hunt.reverse_bonus")
        refractory = Double(b.f("hunt.reverse_refractory"))
        accelMin = b.f("hunt.accel_min")
        accelGain = b.f("hunt.accel_gain_per_frame")
        trigger = max(b.f("hunt.trigger"), 0.001)
        resetTo = b.f("hunt.reset")
        crouchTime = Double(b.f("hunt.crouch"))
        recoverTime = Double(b.f("hunt.recover"))
        attack = max(b.f("hunt.attack"), 0.001)
        crouchSquash = b.f("hunt.squash")
        lean = b.f("hunt.lean")
        bodyLean = b.f("hunt.body_lean")
        pawReach = b.f("hunt.paw_reach")
        wiggleHz = b.f("hunt.wiggle_hz")
        wiggleAmp = b.f("hunt.wiggle_amp")
    }

    func update(_ ctx: TickContext) -> ModuleOutput {
        guard tunedAtlas() != nil else { return .none }
        let stage = CatStage.shared
        let now = CFAbsoluteTimeGetCurrent()
        elapsed += ctx.dt

        let v = ctx.cursorVelocity
        let speed = hypot(v.x, v.y)

        // Direct manipulation wins outright. A cat being dragged or stretched is not
        // also stalking, and the accumulator is emptied rather than frozen so that
        // letting go does not immediately fire whatever built up during the drag.
        let manipulated = stage.state == .dragging || stage.state == .stretching
        if manipulated {
            energy = 0
            phase = .idle
        } else {
            accumulate(now: now, dt: ctx.dt, v: v, speed: speed)
        }
        lastVelocity = v

        // --- pounce state machine -------------------------------------------
        if phase == .idle && energy >= trigger {
            phase = .crouch
            phaseUntil = now + crouchTime
            // Not zero: a cat that has just been teased is easier to tease again,
            // and emptying it would make a second pounce need the full build-up.
            energy = trigger * resetTo
        }
        switch phase {
        case .crouch: if now >= phaseUntil { phase = .recover; phaseUntil = now + recoverTime }
        case .recover: if now >= phaseUntil { phase = .idle }
        case .idle: break
        }

        // Ramp in over `attack`; ramp out over `recover`, which is a time constant of
        // a third of it so the pose has visibly finished when the phase does.
        let tau = phase == .crouch ? attack : CGFloat(recoverTime) / 3
        pose += ((phase == .crouch ? 1 : 0) - pose) * (1 - exp(-ctx.dt / max(tau, 0.001)))

        stage.metric("hunt.e", energy)
        stage.metric("hunt.spd", speed)
        stage.metric("hunt.rev", reversals)

        guard pose > 0.002 else { return .none }

        // --- the crouch ------------------------------------------------------
        var out = ModuleOutput()
        let dist = max(hypot(ctx.cursor.x, ctx.cursor.y), 0.0001)
        let dir = CGPoint(x: ctx.cursor.x / dist, y: ctx.cursor.y / dist)
        // The haunch wiggle every cat does before it commits.
        let wiggle = sin(elapsed * 2 * .pi * wiggleHz) * wiggleAmp * pose

        stage.headOffset.x += dir.x * lean * pose + wiggle
        stage.headOffset.y += dir.y * lean * 0.6 * pose
        stage.pawOffsetL.x += dir.x * pawReach * pose
        stage.pawOffsetR.x += dir.x * pawReach * pose
        stage.tailOffset.x -= dir.x * pawReach * pose

        out.offset.x = dir.x * bodyLean * pose + wiggle * 0.5
        // Squash alone lowers the cat: the rig lifts the body by (scale - 1), so a
        // scale below 1 drops it. Getting low IS the crouch.
        out.squash = 1 - (1 - crouchSquash) * pose
        // The state lasts exactly as long as the crouch and the return; the last few
        // frames of the pose settling are not "hunting" any more.
        if phase != .idle { out.state = .hunting }
        return out
    }

    private func accumulate(now: CFAbsoluteTime, dt: CGFloat, v: CGPoint, speed: CGFloat) {
        // Decay is quoted per frame at a nominal 60fps and normalised by dt, so the
        // accumulator has the same half-life whatever rate the tick actually runs at.
        energy *= pow(decayPerFrame, dt * 60)

        // The gate term. Bounded by construction: sustained excess speed E settles
        // at E * gain / (1 - decay), which is deliberately under the trigger.
        if speed > speedMin {
            energy += (speed - speedMin) * speedGain * dt * 60
        }

        // The term that actually matters. Compared against the heading from
        // `reverseLag` ago, and only counted once per refractory window — a single
        // reversal spans several 120Hz frames and would otherwise be paid for twice.
        trail.append((now, v))
        if trail.count > trailCap { trail.removeFirst(trail.count - trailCap) }
        if let old = trail.last(where: { now - $0.t >= reverseLag }) {
            let oldSpeed = hypot(old.v.x, old.v.y)
            if min(speed, oldSpeed) > reverseSpeed, now - lastReverse > refractory {
                let dot = (v.x * old.v.x + v.y * old.v.y) / (speed * oldSpeed)
                if dot < reverseDot {
                    energy += reverseBonus
                    lastReverse = now
                    reversals += 1
                }
            }
        }

        // A smaller nudge for sheer violence of movement, which catches a flick that
        // is over before it has time to reverse.
        if dt > 0 {
            let accel = hypot(v.x - lastVelocity.x, v.y - lastVelocity.y) / dt
            if accel > accelMin {
                energy += (accel - accelMin) * accelGain * dt * 60
            }
        }
    }
}

File: windows/LoafCat/Modules/HuntModule.cs
namespace LoafCat.Modules;

/// Stalking a cursor that is being wiggled like a cat toy.
///
/// The thing this deliberately is NOT is a speed threshold. A speed threshold fires on
/// every ordinary sweep across a wide display, which is most cursor movement, and then
/// the cat is permanently crouched. What actually reads as "prey" is *changing
/// direction*: a real cat ignores a mouse running in a straight line past it and
/// fixates on one that jinks.
///
/// So this is an energy accumulator whose reversal term dwarfs its speed term. The
/// speed term exists only to gate — its ceiling, `gain * excess / (1 - decay)`, is
/// tuned to sit below the trigger for any plausible straight sweep, so speed alone can
/// never pounce the cat. Reversals are what carry it over.
public sealed class HuntModule : ICatModule, IAtlasTuned
{
    public string Id => "hunt";
    public int TunedGeneration { get; set; } = -1;

    // --- constants, every one of them from cat.json -------------------------
    private double _decayPerFrame = 1;
    private double _speedMin;
    private double _speedGain;
    private double _reverseLag;
    private double _reverseDot;
    private double _reverseSpeed;
    private double _reverseBonus;
    private double _refractory;
    private double _accelMin;
    private double _accelGain;
    private double _trigger = 1;
    private double _resetTo;
    private double _crouchTime;
    private double _recoverTime;
    private double _attack = 0.1;
    private double _crouchSquash = 1;
    private double _lean;
    private double _bodyLean;
    private double _pawReach;
    private double _wiggleHz;
    private double _wiggleAmp;

    // --- state --------------------------------------------------------------
    private enum Phase { Idle, Crouch, Recover }
    private Phase _phase = Phase.Idle;
    private double _phaseUntil;
    private double _energy;
    private double _pose;
    private double _elapsed;
    private Pt _lastVelocity = Pt.Zero;
    private double _lastReverse;
    private double _reversals;

    /// A short trail of velocities, so a reversal is measured against where the cursor
    /// was heading a moment ago rather than one frame ago. At 120Hz the frame-to-frame
    /// angle is mostly EMA noise; over ~60ms it is the gesture.
    private readonly List<(double T, Pt V)> _trail = [];
    private const int TrailCap = 48;

    public void Retune(Atlas atlas)
    {
        var b = atlas.Behaviour;
        _decayPerFrame = b.F("hunt.decay_per_frame");
        _speedMin = b.F("hunt.speed_min");
        _speedGain = b.F("hunt.speed_gain_per_frame");
        _reverseLag = b.F("hunt.reverse_lag");
        _reverseDot = b.F("hunt.reverse_dot");
        _reverseSpeed = b.F("hunt.reverse_speed");
        _reverseBonus = b.F("hunt.reverse_bonus");
        _refractory = b.F("hunt.reverse_refractory");
        _accelMin = b.F("hunt.accel_min");
        _accelGain = b.F("hunt.accel_gain_per_frame");
        _trigger = Math.Max(b.F("hunt.trigger"), 0.001);
        _resetTo = b.F("hunt.reset");
        _crouchTime = b.F("hunt.crouch");
        _recoverTime = b.F("hunt.recover");
        _attack = Math.Max(b.F("hunt.attack"), 0.001);
        _crouchSquash = b.F("hunt.squash");
        _lean = b.F("hunt.lean");
        _bodyLean = b.F("hunt.body_lean");
        _pawReach = b.F("hunt.paw_reach");
        _wiggleHz = b.F("hunt.wiggle_hz");
        _wiggleAmp = b.F("hunt.wiggle_amp");
    }

    public ModuleOutput Update(in TickContext ctx)
    {
        if (this.TunedAtlas() is null) return ModuleOutput.None;
        var stage = CatStage.Shared;
        double now = Clock.Now;
        _elapsed += ctx.Dt;

        var v = ctx.CursorVelocity;
        double speed = MathX.Hypot(v.X, v.Y);

        // Direct manipulation wins outright. A cat being dragged or stretched is not
        // also stalking, and the accumulator is emptied rather than frozen so that
        // letting go does not immediately fire whatever built up during the drag.
        bool manipulated = stage.State is CatState.Dragging or CatState.Stretching;
        if (manipulated)
        {
            _energy = 0;
            _phase = Phase.Idle;
        }
        else
        {
            Accumulate(now, ctx.Dt, v, speed);
        }
        _lastVelocity = v;

        // --- pounce state machine -------------------------------------------
        if (_phase == Phase.Idle && _energy >= _trigger)
        {
            _phase = Phase.Crouch;
            _phaseUntil = now + _crouchTime;
            // Not zero: a cat that has just been teased is easier to tease again, and
            // emptying it would make a second pounce need the full build-up.
            _energy = _trigger * _resetTo;
        }
        switch (_phase)
        {
            case Phase.Crouch:
                if (now >= _phaseUntil) { _phase = Phase.Recover; _phaseUntil = now + _recoverTime; }
                break;
            case Phase.Recover:
                if (now >= _phaseUntil) _phase = Phase.Idle;
                break;
        }

        // Ramp in over `attack`; ramp out over `recover`, which is a time constant of a
        // third of it so the pose has visibly finished when the phase does.
        double tau = _phase == Phase.Crouch ? _attack : _recoverTime / 3;
        _pose += ((_phase == Phase.Crouch ? 1 : 0) - _pose)
                 * (1 - Math.Exp(-ctx.Dt / Math.Max(tau, 0.001)));

        stage.Metric("hunt.e", _energy);
        stage.Metric("hunt.spd", speed);
        stage.Metric("hunt.rev", _reversals);

        if (_pose <= 0.002) return ModuleOutput.None;

        // --- the crouch ------------------------------------------------------
        var outv = new ModuleOutput();
        double dist = Math.Max(MathX.Hypot(ctx.Cursor.X, ctx.Cursor.Y), 0.0001);
        var dir = new Pt(ctx.Cursor.X / dist, ctx.Cursor.Y / dist);
        // The haunch wiggle every cat does before it commits.
        double wiggle = Math.Sin(_elapsed * 2 * Math.PI * _wiggleHz) * _wiggleAmp * _pose;

        stage.HeadOffset.X += dir.X * _lean * _pose + wiggle;
        stage.HeadOffset.Y += dir.Y * _lean * 0.6 * _pose;
        stage.PawOffsetL.X += dir.X * _pawReach * _pose;
        stage.PawOffsetR.X += dir.X * _pawReach * _pose;
        stage.TailOffset.X -= dir.X * _pawReach * _pose;

        outv.Offset.X = dir.X * _bodyLean * _pose + wiggle * 0.5;
        // Squash alone lowers the cat: the rig lifts the body by (scale - 1), so a
        // scale below 1 drops it. Getting low IS the crouch.
        outv.Squash = 1 - (1 - _crouchSquash) * _pose;
        // The state lasts exactly as long as the crouch and the return; the last few
        // frames of the pose settling are not "hunting" any more.
        if (_phase != Phase.Idle) outv.State = CatState.Hunting;
        return outv;
    }

    private void Accumulate(double now, double dt, Pt v, double speed)
    {
        // Decay is quoted per frame at a nominal 60fps and normalised by dt, so the
        // accumulator has the same half-life whatever rate the tick actually runs at.
        _energy *= Math.Pow(_decayPerFrame, dt * 60);

        // The gate term. Bounded by construction: sustained excess speed E settles at
        // E * gain / (1 - decay), which is deliberately under the trigger.
        if (speed > _speedMin)
        {
            _energy += (speed - _speedMin) * _speedGain * dt * 60;
        }

        // The term that actually matters. Compared against the heading from
        // `reverseLag` ago, and only counted once per refractory window — a single
        // reversal spans several 120Hz frames and would otherwise be paid for twice.
        _trail.Add((now, v));
        if (_trail.Count > TrailCap) _trail.RemoveRange(0, _trail.Count - TrailCap);

        // The most recent sample that is already at least `reverseLag` old, which is
        // Swift's `last(where:)` over a list ordered oldest-first.
        for (int i = _trail.Count - 1; i >= 0; i--)
        {
            if (now - _trail[i].T < _reverseLag) continue;
            var old = _trail[i];
            double oldSpeed = MathX.Hypot(old.V.X, old.V.Y);
            if (Math.Min(speed, oldSpeed) > _reverseSpeed && now - _lastReverse > _refractory)
            {
                double dot = (v.X * old.V.X + v.Y * old.V.Y) / (speed * oldSpeed);
                if (dot < _reverseDot)
                {
                    _energy += _reverseBonus;
                    _lastReverse = now;
                    _reversals += 1;
                }
            }
            break;
        }

        // A smaller nudge for sheer violence of movement, which catches a flick that is
        // over before it has time to reverse.
        if (dt > 0)
        {
            double accel = MathX.Hypot(v.X - _lastVelocity.X, v.Y - _lastVelocity.Y) / dt;
            if (accel > _accelMin)
            {
                _energy += (accel - _accelMin) * _accelGain * dt * 60;
            }
        }
    }
}
